storage/file_persister.go

```diff
@@ -43,3 +48,142 @@ func (l *LocalFilePersister) Persist(path string, data io.Reader) (err error) {
+// retrieve one pre-signed URL. The pre-signed url is used to upload the file
+// to the remote location.
+type RemoteFilePersister struct {
+	preSignedURLGetterURL string
+	headers               map[string]string
+	basePath              string
+}
+
+// Persist will upload the contents of data to a remote location.
+func (r *RemoteFilePersister) Persist(ctx context.Context, path string, data io.Reader) (err error) {
+	pURL, err := r.getPreSignedURL(ctx, path)
+	if err != nil {
+		return fmt.Errorf("get presigned url: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPut, pURL, data)
+	if err != nil {
+		return fmt.Errorf("create upload request: %w", err)
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("perform upload request: %w", err)
+	}
+	defer func() {
+		if _, cerr := io.Copy(io.Discard, resp.Body); cerr != nil && err == nil {
+			err = fmt.Errorf("drain upload response body: %w", cerr)
+		}
+
+		if cerr := resp.Body.Close(); cerr != nil && err == nil {
+			err = fmt.Errorf("close upload response body: %w", cerr)
+		}
+	}()
+
+	if err := checkStatusCode(resp); err != nil {
+		return fmt.Errorf("upload status code out of range: %w", err)
+	}
+
+	return err
+}
+
+func checkStatusCode(resp *http.Response) error {
+	if resp.StatusCode == 429 {
+		return fmt.Errorf("request is being rate limited")
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return fmt.Errorf("status code is %d", resp.StatusCode)
+	}
+
+	return nil
+}
+
+// getPreSignedURL will retrieve the presigned url for the current file.
+func (r *RemoteFilePersister) getPreSignedURL(ctx context.Context, path string) (string, error) {
+	b, err := buildPresignedRequestBody(r.basePath, path)
+	if err != nil {
+		return "", fmt.Errorf("build request body: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, r.preSignedURLGetterURL, b)
+	if err != nil {
+		return "", fmt.Errorf("create request: %w", err)
+	}
+
+	for k, v := range r.headers {
+		req.Header.Add(k, v)
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("perform request: %w", err)
+	}
+	defer func() {
+		if cerr := resp.Body.Close(); cerr != nil && err == nil {
+			err = fmt.Errorf("close response body: %w", cerr)
+		}
+	}()
+
+	if err := checkStatusCode(resp); err != nil {
+		return "", fmt.Errorf("status code out of range: %w", err)
+	}
+
+	return readResponseBody(resp)
+}
+
+func buildPresignedRequestBody(basePath, path string) (io.Reader, error) {
+	b := struct {
+		Service string `json:"service"`
+		Files   []struct {
+			Name string `json:"name"`
+		} `json:"files"`
+	}{
+		Service: "aws_s3",
+		Files: []struct {
+			Name string `json:"name"`
+		}{
+			{
+				Name: filepath.Join(basePath, path),
+			},
+		},
+	}
+
+	bb, err := json.Marshal(b)
+	if err != nil {
+		return nil, fmt.Errorf("marshal request body: %w", err)
+	}
+
+	return bytes.NewReader(bb), nil
```

Comment:
Not sure i understand why it's unusual, do you have an example why? In this scenario the response is being directly used in the request which takes in an `io.Reader`.

Reply:
[Here](https://pkg.go.dev/io#LimitReader), for example, it makes sense because it takes a `Reader` and changes its behavior with a new one. Returning a `Reader` is a must in this case.

Reply:
I can see that there's no point in this particular case for `buildPresignedRequestBody` since the body is tiny. Resolved in a8991fdfbc46e2b939dc756a26d4498091b87dd3.

Reply:
Yep. A better way would be returning a concrete `bytes.Reader` if the body was large, or just joining those functions together.